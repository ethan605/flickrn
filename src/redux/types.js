export const PREFIXES = {
  FLICKR: 'flickr/',
  SHARED: 'shared/',
};

export const FLICKR = {
  CLEAR_PHOTO_INFO: 'flickr/CLEAR_PHOTO_INFO',
  GET_PHOTO_INFO: 'flickr/GET_PHOTO_INFO',
  SEARCH_PHOTOS: 'flickr/SEARCH_PHOTOS',
};

export const SHARED = {
};
