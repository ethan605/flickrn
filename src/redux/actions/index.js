import * as flickr from './flickr';

/* eslint-disable import/prefer-default-export */

export { flickr };
