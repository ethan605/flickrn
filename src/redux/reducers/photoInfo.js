import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import PhotoInfo from 'src/models/PhotoInfo';

const INITIAL_STATE = {
  data: undefined,
  errorMessage: undefined,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photoInfoReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.CLEAR_PHOTO_INFO)
    return INITIAL_STATE;

  if (type === FLICKR.GET_PHOTO_INFO + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO + SUCCESS) {
    const { payload: { photo } } = action;
    if (_.isEmpty(photo)) return state;

    return { ...state, data: PhotoInfo.build(photo), loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO)
    return { ...state, data: undefined, errorMessage: undefined, loading: true };

  return state;
}
