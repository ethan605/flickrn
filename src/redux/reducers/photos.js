import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import Photo from 'src/models/Photo';

const INITIAL_STATE = {
  currentPage: 1,
  data: [],
  errorMessage: undefined,
  hasMore: true,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photosReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.SEARCH_PHOTOS + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, hasMore: false, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS + SUCCESS) {
    const { payload: { photos } } = action;
    if (_.isEmpty(photos)) return state;

    const { page, pages, photo } = photos;
    const hasMore = page < pages;

    const { data: currentPhotos } = state;
    const newPhotos = Photo.buildArray(photo);

    const data = page === 1 ? newPhotos : [...currentPhotos, ...newPhotos];
    return { ...state, hasMore, data, currentPage: page, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS)
    return { ...state, errorMessage: undefined, loading: true };

  return state;
}
