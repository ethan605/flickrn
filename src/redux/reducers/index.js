import photoInfo from './photoInfo';
import photos from './photos';

export { photoInfo, photos };
