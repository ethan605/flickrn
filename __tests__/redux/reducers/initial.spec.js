import { photoInfo, photos } from 'src/redux/reducers';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';

describe('redux/reducers/initial', () => {
  const mockState = buildMockState();

  test('photoInfo', () => {
    const { photoInfo: expectedState } = mockState;

    expect(photoInfo(undefined, { type: 'DUMMY_TYPE' })).toEqual(expectedState);
    expect(photoInfo(undefined, { type: 'DUMMY_TYPE' })).toMatchSnapshot();
  });

  test('photos', () => {
    const { photos: expectedState } = mockState;

    expect(photos(undefined, { type: 'DUMMY_TYPE' })).toEqual(expectedState);
    expect(photos(undefined, { type: 'DUMMY_TYPE' })).toMatchSnapshot();
  });
});
