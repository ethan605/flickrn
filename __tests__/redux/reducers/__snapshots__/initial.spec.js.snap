// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/initial photoInfo 1`] = `
Object {
  "data": undefined,
  "errorMessage": undefined,
  "loading": false,
}
`;

exports[`redux/reducers/initial photos 1`] = `
Object {
  "currentPage": 1,
  "data": Array [],
  "errorMessage": undefined,
  "hasMore": true,
  "loading": false,
}
`;
